pkg/gfx/bufferarray.go

```diff
@@ -0,0 +1,97 @@
+package gfx
+
+import (
+	"unsafe"
+
+	"github.com/go-gl/gl/v2.1/gl"
+	"github.com/gregjohnson2017/tabula-editor/pkg/log"
+)
+
+// BufferArray is used to efficiently render shapes with OpenGL.
+type BufferArray struct {
+	vaoID      uint32
+	vboID      uint32
+	mode       uint32
+	vertSize   int32
+	numAttribs uint32
+	data       []float32
+}
+
+// NewBufferArray creates the structure necessary for efficiently rendering
+// shapes in OpenGL. It configures a VAO & VBO pair with a specified mode and
+// vertex layout. Example mode: gl.TRIANGLES. Example vertex layout: (x,y,z,
+// s,t) -> layout = (3, 2).
+func NewBufferArray(mode uint32, layout []int32) *BufferArray {
+	var vaoID, vboID uint32
+	gl.GenVertexArrays(1, &vaoID)
+	gl.GenBuffers(1, &vboID)
+	configureVAO(vaoID, vboID, layout)
+	var vertSize int32
+	for _, s := range layout {
+		vertSize += s
+	}
+	return &BufferArray{vaoID, vboID, mode, vertSize, uint32(len(layout)), nil}
+}
+
+// configureVAO configures a VAO & VBO pair with a specified vertex layout
+// example vertex layout: (x,y,z, s,t) -> layout = (3, 2)
+func configureVAO(vaoID uint32, vboID uint32, layout []int32) {
+	gl.BindBuffer(gl.ARRAY_BUFFER, vboID)
+	gl.BindVertexArray(vaoID)
+
+	var vertexSize int32
+	for i := 0; i < len(layout); i++ {
+		vertexSize += layout[i]
+	}
```

Comment:
This loop can be replaced with a range, like in the constructor. Or maybe we pass it in to the function, since we already have it. Or merge with constructor.

Reply:
done